lib/sqlvalue/src/lib.rs

```diff
@@ -0,0 +1,204 @@
+// SqlValue is a dynamically-typed object that holds a value that at
+// runtime can appear in a SQL program.  These values are not used
+// during computations, but they are used to display values.  The
+// Tuple* types are used for computations, and they are converted
+// to SqlRow objects when they need to be displayed.
+
+use ordered_float::OrderedFloat;
+
+pub enum SqlValue {
+    Int(i32),
+    Str(String),
+    Flt(f32),
+    Dbl(f64),
+    Bool(bool),
+
+    OptInt(Option<i32>),
+    OptStr(Option<String>),
+    OptFlt(Option<f32>),
+    OptDbl(Option<f64>),
+    OptBool(Option<bool>),
+}
+
+impl From<i32> for SqlValue {
+    fn from(value: i32) -> Self {
+        SqlValue::Int(value)
+    }
+}
+
+impl From<bool> for SqlValue {
+    fn from(value: bool) -> Self {
+        SqlValue::Bool(value)
+    }
+}
+
+impl From<f32> for SqlValue {
+    fn from(value: f32) -> Self {
+        SqlValue::Flt(value)
+    }
+}
+
+impl From<f64> for SqlValue {
+    fn from(value: f64) -> Self {
+        SqlValue::Dbl(value)
+    }
+}
+
+impl From<OrderedFloat<f32>> for SqlValue {
+    fn from(value: OrderedFloat<f32>) -> Self {
+        SqlValue::Flt(value.into())
+    }
+}
+
+impl From<OrderedFloat<f64>> for SqlValue {
+    fn from(value: OrderedFloat<f64>) -> Self {
+        SqlValue::Dbl(value.into())
+    }
+}
+
+impl From<String> for SqlValue {
+    fn from(value: String) -> Self {
+        SqlValue::Str(value)
+    }
+}
+
+impl From<Option<i32>> for SqlValue {
+    fn from(value: Option<i32>) -> Self {
+        SqlValue::OptInt(value)
+    }
+}
+
+impl From<Option<bool>> for SqlValue {
+    fn from(value: Option<bool>) -> Self {
+        SqlValue::OptBool(value)
+    }
+}
+
+impl From<Option<f32>> for SqlValue {
+    fn from(value: Option<f32>) -> Self {
+        SqlValue::OptFlt(value)
+    }
+}
+
+impl From<Option<f64>> for SqlValue {
+    fn from(value: Option<f64>) -> Self {
+        SqlValue::OptDbl(value)
+    }
+}
+
+impl From<Option<OrderedFloat<f32>>> for SqlValue {
+    fn from(value: Option<OrderedFloat<f32>>) -> Self {
+        match value {
+            None => SqlValue::OptFlt(None),
+            Some(OrderedFloat(x)) => SqlValue::OptFlt(Some(x)),
+        }
+    }
+}
+
+impl From<Option<OrderedFloat<f64>>> for SqlValue {
+    fn from(value: Option<OrderedFloat<f64>>) -> Self {
+        match value {
+            None => SqlValue::OptDbl(None),
+            Some(OrderedFloat(x)) => SqlValue::OptDbl(Some(x)),
+        }
+    }
+}
+
+impl From<Option<String>> for SqlValue {
+    fn from(value: Option<String>) -> Self {
+        SqlValue::OptStr(value)
+    }
+}
+
+pub struct SqlRow {
+    values: Vec<SqlValue>,
+}
+
+pub trait ToSqlRow {
+    fn to_row(&self) -> SqlRow;
+}
+
+impl SqlRow {
+    pub fn new() -> Self {
+        SqlRow { values: Vec::default() }
+    }
+
+    // Output the SqlRow value in the format expected by the tests
```

Comment:
here if you use /// instead of // it will go in rust docs

Reply:
Also might want to document the fact that this panics in case format.len() is not self.values.len(), e.g., smth like 

```
/// # Panics
/// If `format.len() != self.values.len()`
```

in rustdoc syntax is what's usually done